src/effects/backends/builtin/beatgrindeffect.cpp

```diff
@@ -0,0 +1,192 @@
+    triplet->setRange(0, 0, 1);
+
+    return pManifest;
+}
+
+void BeatGrindEffect::loadEngineEffectParameters(
+        const QMap<QString, EngineEffectParameterPointer>& parameters) {
+    m_pDelayParameter = parameters.value("delay_time");
+    m_pQuantizeParameter = parameters.value("quantize");
+    m_pSendParameter = parameters.value("send_amount");
+    m_pFeedbackParameter = parameters.value("feedback_amount");
+    m_pTripletParameter = parameters.value("triplet");
+}
+
+void BeatGrindEffect::processChannel(
+        BeatGrindState* pGroupState,
+        const CSAMPLE* pInput,
+        CSAMPLE* pOutput,
+        const mixxx::EngineParameters& bufferParameters,
+        const EffectEnableState enableState,
+        const GroupFeatureState& groupFeatures) {
+    BeatGrindState& gs = *pGroupState;
+    // The minimum of the parameter is zero so the exact center of the knob is 1 beat.
+
+    double send_current = m_pSendParameter->value();
+    double feedback_current = m_pFeedbackParameter->value();
+    double period = m_pDelayParameter->value();
+
+    if (enableState == EffectEnableState::Enabling) {
+        gs.write_position = 0;
+        gs.read_position = 0;
+        gs.isRecording = true;
+    }
+
+    int delay_frames;
+    if (groupFeatures.has_beat_length_sec) {
+        // period is a number of beats
+        if (m_pQuantizeParameter->toBool()) {
+            period = std::max(roundToFraction(period, 4), 1/8.0);
+            if (m_pTripletParameter->toBool()) {
+                period /= 3.0;
+            }
+        } else if (period < 1/8.0) {
+            period = 1/8.0;
+        }
+        delay_frames = period * groupFeatures.beat_length_sec * bufferParameters.sampleRate();
+    } else {
+        // period is a number of seconds
+        period = std::max(period, 1/8.0);
+        delay_frames = period * bufferParameters.sampleRate();
+    }
+    VERIFY_OR_DEBUG_ASSERT(delay_frames > 0) {
+        delay_frames = 1;
+    }
+
+    int delay_samples = delay_frames * bufferParameters.channelCount();
+    VERIFY_OR_DEBUG_ASSERT(delay_samples <= gs.delay_buf.size()) {
+        delay_samples = gs.delay_buf.size();
+    }
+
+    RampingValue<CSAMPLE_GAIN> send(send_current, gs.prev_send,
+                                    bufferParameters.framesPerBuffer());
+    // Feedback the delay buffer and then add the new input.
+
+    RampingValue<CSAMPLE_GAIN> feedback(feedback_current, gs.prev_feedback,
+                                        bufferParameters.framesPerBuffer());
+    
+    for (unsigned int frame = 0;
+            frame < bufferParameters.samplesPerBuffer();
+            frame += bufferParameters.channelCount()) {
+        CSAMPLE_GAIN send_ramped = send.getNext();
+        CSAMPLE_GAIN feedback_ramped = feedback.getNext();    
+
+        for (int channel = 0; channel < bufferParameters.channelCount(); channel++) {
+            if (gs.isRecording) {
+                if (gs.write_position >= gs.delay_buf.size()) {
```

Comment:
VERIFY_OR_DEBUG_ASSERT? I don't think it should be possible for the time parameter to reach a value bigger than the preallocated delay buffer.

Reply:
No, it shouldn't. But it needs to reach the value that is equal to, to stop writing to the buffer.
The if/else here is enough, it will never write past the size of the buffer, changing it to VERIFY_OR_DEBUG_ASSERT would require another an extra if-else condition

Reply:
okay